### PR TITLE
Fix host port allocation

### PR DESCRIPTION
**Pull request by tbachman**
The allocation of host ports for the external gateway
was broken.

This closes #143

Signed-off-by: Thomas Bachman tbachman@yahoo.com

### CODE REVIEW

File: networking_cisco/plugins/cisco/device_manager/plugging_drivers/aci_vlan_trunking_driver.py [hunk 1]
```diff
@@ -263,11 +263,18 @@ def allocate_hosting_port(self, context, router_id, port_db, network_type,
         # If this is a router interface, the VLAN comes from APIC.
         # If it's the gateway, the VLAN comes from the segment ID
         if port_db.get('device_owner') == DEVICE_OWNER_ROUTER_GW:
-            return super(AciVLANTrunkingPlugDriver,
-                         self).allocate_hosting_port(
-                             context, router_id,
-                             port_db, network_type, hosting_device_id
-                         )
+            ext_dict, net = self._get_external_network_dict(context, port_db)
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/device_manager/plugging_drivers/aci_vlan_trunking_driver.py:257]: should we do some sanity check to make sure (ext_dict, net) is available before we use them? For example if user forgot to provide the needed section/parameters in the config file.

File: networking_cisco/tests/unit/cisco/device_manager/test_aci_vlan_trunking_driver.py [hunk 2]
```diff
@@ -453,3 +452,56 @@ def _return_mocked_net(self, args):
                     self.plugging_driver.extend_hosting_port_info(ctx,
                         fake_port_db_obj, hosting_device, hosting_info)
                     self.assertEqual(hosting_info['segmentation_id'], 3003)
+
+    def test_allocate_hosting_port_info_adds_segment_id(self):
+        self.plugging_driver._default_ext_dict = {
+            'gateway_ip': '1.103.2.1',
```
Comment (kentwu1) [inline networking_cisco/tests/unit/cisco/device_manager/test_aci_vlan_trunking_driver.py]: the naming of these parameters are still not consistent with the naming on the neutron side.

File: networking_cisco/plugins/cisco/device_manager/plugging_drivers/aci_vlan_trunking_driver.py (hunk 1 above)
Comment (tbachman) [inline networking_cisco/plugins/cisco/device_manager/plugging_drivers/aci_vlan_trunking_driver.py:257]: There's a default set that's provided. If the user provided the correct network, but didn't provide the correct information for that network, then an exception should be thrown.

**Comment by tbachman** [comment]
Rebased the patch, which picks up the corrected config file options from #33

File: networking_cisco/plugins/cisco/device_manager/plugging_drivers/aci_vlan_trunking_driver.py [hunk 3]
```diff
@@ -254,11 +254,18 @@ def allocate_hosting_port(self, context, router_id, port_db, network_type,
...
+            # If an OpFlex network is used on the external network,
+            # the actual segment ID comes from the confgi file
+            if net.get('provider:network_type') == 'opflex':
+                if ext_dict.get('segmentation_id'):
+                    return {'allocated_port_id': port_db.id,
+                            'allocated_vlan': ext_dict['segmentation_id']}
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/device_manager/plugging_drivers/aci_vlan_trunking_driver.py:263]: as we discussed, if 'segmentation_id' is not present, then a proper exception, error log should be given.

File: networking_cisco/plugins/cisco/device_manager/plugging_drivers/aci_vlan_trunking_driver.py (hunk 1 above)
Comment (kentwu1) [inline networking_cisco/plugins/cisco/device_manager/plugging_drivers/aci_vlan_trunking_driver.py:257]: what happens if user forgot to provide the entire section at all? then we will just use the default set then everything will seem fine however this is probably not what user wants....Maybe we should just get rid of the default then?

Comment (tbachman) [inline networking_cisco/plugins/cisco/device_manager/plugging_drivers/aci_vlan_trunking_driver.py:257]: We want the following behavior:
1) if it's not an opflex network type, call the superclass to get the segmentation ID
2) if it's an opflex network, get the segmentation ID from the config. If it's not present, we should throw an exception.
The current code should be changed to address number 2 (i.e. no exception is being thrown the way that it's currently written).

File: networking_cisco/plugins/cisco/device_manager/plugging_drivers/aci_vlan_trunking_driver.py (hunk 3 above)
Comment (tbachman) [inline networking_cisco/plugins/cisco/device_manager/plugging_drivers/aci_vlan_trunking_driver.py:263]: Yes, see comment above.

**Comment by tbachman** [comment]
Provided update, based on reviewer comments; also removed now-unused code.

File: networking_cisco/plugins/cisco/device_manager/plugging_drivers/aci_vlan_trunking_driver.py [hunk 4]
```diff
@@ -254,11 +254,19 @@ def allocate_hosting_port(self, context, router_id, port_db, network_type,
...
         if port_db.get('device_owner') == DEVICE_OWNER_ROUTER_GW:
+            ext_dict, net = self._get_external_network_dict(context, port_db)
+            # If an OpFlex network is used on the external network,
+            # the actual segment ID comes from the confgi file
+            if net.get('provider:network_type') == 'opflex':
+                if ext_dict.get('segmentation_id'):
+                    return {'allocated_port_id': port_db.id,
+                            'allocated_vlan': ext_dict['segmentation_id']}
+                else:
+                    raise AciDriverConfigMissingSegmentationId(ext_net=net)
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/device_manager/plugging_drivers/aci_vlan_trunking_driver.py:265]: add a test case for this exception then?

**Pull request merged**